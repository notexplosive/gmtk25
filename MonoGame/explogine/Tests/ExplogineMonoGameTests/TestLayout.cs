using ApprovalTests;
using ExplogineCore.Data;
using ExplogineMonoGame.Data;
using ExplogineMonoGame.Layout;
using FluentAssertions;
using Microsoft.Xna.Framework;
using Xunit;

namespace ExplogineMonoGameTests;

public class TestLayout
{
    [Fact]
    public void build_serial_from_real_layout()
    {
        var layout = L.Compute(
            new RectangleF(0, 0, 500, 500),
            new LayoutElementGroup(
                new Style(
                    Orientation.Vertical,
                    Margin: new Vector2(25, 25)),
                new[]
                {
                    L.Group(L.FillHorizontal("title-bar", 40),
                        new Style(
                            Alignment: Alignment.Center),
                        new[]
                        {
                            L.FixedElement("icon", 32, 32),
                            L.Group(L.FillHorizontal(32),
                                new Style(
                                    Alignment: Alignment.CenterRight,
                                    PaddingBetweenElements: 3),
                                new[]
                                {
                                    L.FillHorizontal("title", 32),
                                    L.FixedElement("minimize-button", 32, 32),
                                    L.FixedElement("fullscreen-button", 32, 32),
                                    L.FixedElement("close-button", 32, 32)
                                })
                        }),
                    L.FillBoth("body")
                }));

        Approvals.Verify(L.ToJson(layout));
    }

    [Fact]
    public void serialize_and_back()
    {
        var layout = L.Compute(
            new RectangleF(0, 0, 500, 500),
            new LayoutElementGroup(
                new Style(
                    Orientation.Vertical,
                    Margin: new Vector2(25, 25)),
                new[]
                {
                    L.Group(L.FillHorizontal("title-bar", 40),
                        new Style(
                            Alignment: Alignment.Center),
                        new[]
                        {
                            L.FixedElement("icon", 32, 32),
                            L.Group(L.FillHorizontal(32),
                                new Style(
                                    Alignment: Alignment.CenterRight,
                                    PaddingBetweenElements: 3),
                                new[]
                                {
                                    L.FillHorizontal("title", 32),
                                    L.FixedElement("minimize-button", 32, 32),
                                    L.FixedElement("fullscreen-button", 32, 32),
                                    L.FixedElement("close-button", 32, 32)
                                })
                        }),
                    L.FillBoth("body")
                }));

        var serialized = L.ToJson(layout);
        var deserialized = L.FromJson(new RectangleF(0, 0, 500, 500), serialized);

        L.ToJson(deserialized).Should().Be(serialized);

    }
}
