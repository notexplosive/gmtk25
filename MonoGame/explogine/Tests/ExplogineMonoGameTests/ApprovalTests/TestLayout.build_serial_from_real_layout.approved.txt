{
  "Settings": {
    "Orientation": 1,
    "Margin": [
      25.0,
      25.0
    ],
    "Alignment": "Left Top",
    "Padding": 0
  },
  "Elements": [
    {
      "Name": "title-bar",
      "Size": [
        "fill",
        "40"
      ],
      "SubGroup": {
        "Settings": {
          "Orientation": 0,
          "Margin": [
            0.0,
            0.0
          ],
          "Alignment": "Center Center",
          "Padding": 0
        },
        "Elements": [
          {
            "Name": "icon",
            "Size": [
              "32",
              "32"
            ],
            "SubGroup": null
          },
          {
            "Name": null,
            "Size": [
              "fill",
              "32"
            ],
            "SubGroup": {
              "Settings": {
                "Orientation": 0,
                "Margin": [
                  0.0,
                  0.0
                ],
                "Alignment": "Right Center",
                "Padding": 3
              },
              "Elements": [
                {
                  "Name": "title",
                  "Size": [
                    "fill",
                    "32"
                  ],
                  "SubGroup": null
                },
                {
                  "Name": "minimize-button",
                  "Size": [
                    "32",
                    "32"
                  ],
                  "SubGroup": null
                },
                {
                  "Name": "fullscreen-button",
                  "Size": [
                    "32",
                    "32"
                  ],
                  "SubGroup": null
                },
                {
                  "Name": "close-button",
                  "Size": [
                    "32",
                    "32"
                  ],
                  "SubGroup": null
                }
              ]
            }
          }
        ]
      }
    },
    {
      "Name": "body",
      "Size": [
        "fill",
        "fill"
      ],
      "SubGroup": null
    }
  ]
}